internal/confmapprovider/discovery/config.go

```diff
@@ -0,0 +1,356 @@
+			}
+			return loadEntry(typeReceiver, path, c.Receivers)
+		default:
+			c.logger.Debug(fmt.Sprintf("Disregarding %s", path))
+		}
+		return nil
+	})
+}
+
+func errorF(entryType, path string, err error) error {
+	return fmt.Errorf("failed loading %s from %s: %w", entryType, path, err)
+}
+
+func isServiceEntryPath(path string) bool {
+	return serviceEntryRegex.MatchString(path)
+}
+
+func isExporterEntryPath(path string) bool {
+	return exporterEntryRegex.MatchString(path)
+}
+
+func isExtensionEntryPath(path string) bool {
+	return extensionEntryRegex.MatchString(path)
+}
+
+func isDiscoveryObserverEntryPath(path string) bool {
+	return discoveryObserverEntryRegex.MatchString(path)
+}
+
+func isProcessorEntryPath(path string) bool {
+	return processorEntryRegex.MatchString(path)
+}
+
+func isReceiverEntryPath(path string) bool {
+	return receiverEntryRegex.MatchString(path)
+}
+
+func isReceiverToDiscoverEntryPath(path string) bool {
+	return receiverToDiscoverEntryRegex.MatchString(path)
+}
+
+func loadEntry[K keyType, V entryType](componentType, path string, target map[K]V) error {
+	tmpDest := map[K]V{}
+
+	componentID, err := unmarshalEntry(componentType, path, &tmpDest)
+	noTypeK := noTypeForK(componentID)
+	if err != nil {
+		return tmpDest[noTypeK].ErrorF(path, err)
+	}
+
+	if componentID == noTypeK {
+		return nil
+	}
+
+	if componentType == typeService {
+		// set directly on target and exit
+		for k, v := range tmpDest {
+			target[k] = v
+		}
+		return nil
+	}
+
+	if v, ok := target[componentID]; ok {
+		return v.ErrorF(path, fmt.Errorf("duplicate %q", keyTypeToString(componentID)))
+	}
+	entry := tmpDest[componentID]
+	target[componentID] = entry
+	return nil
+}
+
+func unmarshalEntry[K keyType, V entryType](componentType, path string, dst *map[K]V) (componentID K, err error) {
+	if dst == nil {
+		err = fmt.Errorf("cannot load %s into nil entry", componentType)
+		return
+	}
+	if err = unmarshalYaml(path, dst); err != nil {
+		err = fmt.Errorf("failed unmarshalling component %s: %w", componentType, err)
+		return
+	}
+
+	// service is marshaled as complete item so return as is
+	if componentType == typeService {
+		var s any = typeService
+		return s.(K), nil
+	}
+
+	entry := *dst
+
+	if len(entry) == 0 {
+		// empty or all-comment files are supported but ignored
+		var noTypeK any = noType
+		return noTypeK.(K), nil
+	}
+	var componentIDs []K
+	for k := range entry {
+		componentIDs = append(componentIDs, k)
+	}
+
+	if len(componentIDs) != 1 {
+		err = fmt.Errorf("%q must contain a single mapping of ComponentID to component but contained %v", path, componentIDs)
+		return
+	}
+	return componentIDs[0], nil
+}
+
+func unmarshalYaml(path string, out any) error {
+	contents, err := os.ReadFile(filepath.Clean(path))
+	if err != nil {
+		return fmt.Errorf("failed reading file %q: %w", path, err)
+	}
+
+	if err = yaml.Unmarshal(contents, out); err != nil {
+		return fmt.Errorf("failed parsing %q as yaml: %w", path, err)
```

Comment:
Do we stop walking the directory if one file is bad? Can we test for that and make sure we log and continue?

Reply:
On this first pass I'm not under the impression it's a good idea to try to use a config that has invalid sections (the collector also tends to error out earlier on invalid component config). `WalkDir()` will end on the first non-skipable error. Updated to clear the config maps on error to be clear in this regard and added an invalid item test, but I'm happy to update to be more intuitive.

Reply:
The problem I see often in customer envs is that when we fail, we crash, throw an error such "cannot read the yaml", and the customer is left picking up the pieces.
Given that we now have a few of those files in these directories, it might be good to help customers get by.

Note that in the root config case, the service definition contains the logger severity level - meaning they are cues we log, but might not be seen by the customer.

This is out of scope for this PR, but error resilience - bad inputs, horrible configs - we will need to address that to make our lives easier.

Reply:
I'd hope in the future that we offer something like a full service otto the collector hosts locally that can be used to created a validated config that the collector service uses on a reload. If we fully adopt opamp in product hopefully a local web service would be provided that uses it as well.

Reply:
That's cool. I believe that might be a future epic - make our lives easier when it's debug time.